File: R/code/02_dirichlet_process_mixtures.R
source('code/00_extras.R')
source('code/01_dirichlet_process.R')


# Data simulation from a DPM with Normal model for the data --------------------------
tic_rdpm_data <- function(n, M, m, tau, s, S){
  # TODO: Use any model for the data
  # TODO: Use any centering measure
  
  # Parameter simulation
  pi_vector <- tic_rdp_data(
    n, M, rnormigamma,
    G0_params = list(
      mu = m, lambda = 1 / tau,
      alpha = s / 2, beta = 2 / S
    )
  )
  
  # Number of distinct components
  k <- length(unique(pi_vector[, 1]))
  
  # Data
  y <- apply(
    pi_vector, 1, function(params) rnorm(1, mean = params[1], sd = sqrt(params[2]))
  )
  
  # List with all the values
  return(list(y = y, params = pi_vector, k = k))
}

File: Julia/code/02_dirichlet_process_mixtures.jl
include("01_dirichlet_process.jl")


# Data simulation from a DPM of normals
function tic_rdpm_normal(n::Int64, M, m, tau, s, S)
    """
    Random sample from a Dirichlet Process Mixture of Normals. Returns the component
    parameters, the number of unique components (k) and the simulated data

    n            : sample length
    M            : precision parameter
    m, tau, s, S : parameters for the Normal-Inv-Gamma distribution

    TODO: Use any model for the data
    TODO: Use any centering measure G0
    """

    G0 = NormalInverseGamma(m, tau, s, S)
    pi_vector = tic_rdp_marginal(n, M, G0)
    k = size(unique(pi_vector, dims=1))[1]
    y = map(params -> rand(Normal(params[1], sqrt(params[2]))), eachrow(pi_vector))

    return (y=y, params=pi_vector, k=k)
end
